docs/buffers.md

```diff
@@ -0,0 +1,143 @@
+unsigned int *host_data = (unsigned int *)acl_aligned_malloc(global_work_size*sizeof(unsigned int)*GMEM_DATA_VEC);
+
+// Set up the expectation for a buffer, but not actually allocating memory yet
+cl_mem mem = clCreateBuffer(context, CL_MEM_READ_WRITE, global_work_size*sizeof(unsigned int)*GMEM_DATA_VEC, NULL, &err);
+
+// Enqueue a memory transfer operation
+cl_event event = {};
+err = clEnqueueWriteBuffer(queue, mem, CL_FALSE, 0, global_work_size*sizeof(unsigned int)*GMEM_DATA_VEC, host_data, 0, NULL, &event);
+
+// Informs the kernel that one of its arguments is the buffer
+err = clSetKernelArg(kernel, 0, sizeof(cl_mem), &mem);
+
+// Submit kernel execution to queue
+err = clEnqueueNDRangeKernel(queue, kernel, 2, NULL, &global_work_size_2d[0], &local_work_size_2d[0], 0, NULL, &event);
+
+... // Block on kernel finish and clean up resources
+```
+
+## Summary
+
+* Buffers are not allocated on the device until the kernel executes, the `clEnqueueWriteBuffer` function is called, or `CL_MEM_COPY_HOST_PTR` flag is defined when calling the function `clCreateBuffer`.
+* Buffer handles memory migration between the host and device automatically.
+* SYCL* runtime typically calls OpenCL functions in a different order in comparison to OpenCL's preferred order. Specifically, SYCL* runtime calls functions in the order of `clCreateBuffer` > `clEnqueueWriteBuffer` > `clSetKernelArg` > `clEnqueueNDRangeKernel`. In OpenCL, the optimal calls are `clCreateBuffer` > `clSetKernelArg` > `clEnqueueWriteBuffer` >  `clEnqueueNDRangeKernel`. The difference in order of calls sometimes leads to an unexpected result (eg. extra host to device memory copy), but you can avoid it if the `clSetKernelArg` function does not modify `cl_mem` attributes.
+* The runtime keeps track of which memory addresses are occupied and decides where the next allocation should be. Specifically, it uses a first-fit allocation algorithm.
+* The actual memory allocation is made with MMD calls by passing the device address and size.
+* Device address has a special encoding, which is the same between buffer and USM pointers.
+* Even if you do not explicitly transfer the memory before launching the kernel, the memory is still migrated right before the kernel executes (i.e., `clEnqueueWriteBuffer` is not necessary).
+
+## SYCL Runtime Flow
+
+SYCL* runtime calls the memory operation in the following order, which are explained in the subsections below:
+1. `clCreateBuffer`
+2. `clEnqueueWriteBuffer`
+3. `clSetKernelArg`
+4. `clEnqueueKernel` / Task
+
+
+**NOTE:** SYCL* runtime calls are not the preferred order from the OpenCL runtime perspective. However, SYCL runtime* cannot change the order of API invocations as it needs to handle other vendors' requirements. Not all vendors can get away with setting kernel argument first before enqueuing write buffer (As discussed in one of the [issues](https://github.com/intel/llvm/discussions/4627)).
```

Comment:
OpenCL runtime → FPGA Runtime for OpenCL

Reply:
(As discussed in one of the [issues](https://github.com/intel/llvm/discussions/4627)) → (see [discussion](https://github.com/intel/llvm/discussions/4627))